test/test.sh

```diff
@@ -1,30 +1,35 @@
-#!/bin/bash -e
+#!/bin/sh
+
+set -eux
```

Comment:
add `-o pipefail` to detect upstream pipe command failures

Reply:
/tmp/test.sh: 4: set: Illegal option -o pipefail

Reply:
oh, i see, you changed the interpreter to not use bash. given that this script is only used for testing I think it is fine to use these less-portable features for the sake of easier testing, readability, and bug discovery.  do you have a use case where using bash is not an option?

Reply:
currently, i'm bash-less :) (i use zsh primarly) and for work i use busybox's ash. By forcing bash (and less portable code) i cannot use the test script locally anymore and would have to rely on the CI for output constantly for example.

Reply:
doh!  ok, we certainly don't want to leave anyone in the dust who is inclined to run/update tests! ;)

that said, i think the bash equivalent of `-o pipefail` is an important option we should be enabling.  some quick googling shows there is a hacky way of getting equivalent functionality from posix sh with a utility function: http://cfajohnson.com/shell/cus-faq-2.html#Q11

however, wouldn't it be easier to expect that anyone running tests have bash installed?  what distro are you using?  would it be prohibitive to install bash?  i'd be fine standardizing on zsh too if you think thats better.  portability takes a back seat in this scenario to clear, efficient tests

Reply:
for our current use, I don't think we need it just yet, as if anything fails in the semonitor pipeline, the results should not match and thus the compare's will fail. It would ease debugging :s 
so should we open a new issue for other improvements or you want to pull it in under this issue?

wat we could do is, add 'if sh == bash; set -o pipefail' i suppose?

Reply:
we could add the pipefail option in another PR but then we'll just be back at the same issue with using sh rather than bash.

I want to keep using the more advanced features of bash so if it is prohibitive for you to install bash then please update your PR to use zsh

Reply:
so, then getting back at the issue at hand, why do we need pipefail?

if any of the piiped commands fail, the input to the last function will be wrong, and in no way can the last software produce anything useful and thus the last command, or the check just afterwards will fail horribly anyway

Reply:
consider the case when a valid output (stdout) is empty.  a bad file input or a fatal execution, for example, would be masked unless you check the return code.

you are correct: one intention of adding pipefail is that it will implicitly help test intermediate commands such as unhexlify.  it also helps validate the tests themselves by making the tests more brittle and therefor "fail early" for more bugs.  it brings a lot of value with very little effort

there are plenty of writings supporting this as paradigm:
https://kvz.io/blog/2013/11/21/bash-best-practices/
https://github.com/progrium/bashstyle
"Always use set -eo pipefail. Fail fast and be aware of exit codes."

Reply:
oh i do not argue against pipefail; i'll think of something that makes the subcommand testing even more verbose then

Reply:
i just fixed it and even made it stronger.

giving it some thought, I have to agree with what you say, fail early, fail often and be brittle.
To accomplish this, remove the pipeline entirely. While making the script slower and heavier on disk space (who cares :p) it does improve debugabiltiy of the separate steps even more then a pipefail would. In test scripts, don't pipe! :)

Reply:
yer one tough :cookie: :p

so i've added pipefail and made it into a pipeline again.

in my defense ;) the reason why i want to avoid bash(isms) is that it's non-portable code, and I like/prefer to write portable code generally.

Further more, you are a little contradicting :) you say more tests are good (testing unhexlify); more debugging information is better (set -x + not cleaning up on failure); yet insist on pipefail, which negates these both.

As for the extra files, it's a test script, the files are not gigabytes and are not wasting space (they get removed after the test). But it does give the person running the test valueable feedback as to what went wrong.

Anyway, I submit, you win :) pipefail is in, bash guaranteed to be installed and set as the default shell.